#!/bin/bash

############################################################################################################################
#############
#				SAGPEK: a command-line software to judge the genotypes of investigated loci in a batch of ABI 
#					files generated by Sanger sequencing
#
#				Version:  v1.0
#
#				About: Developed in the Shandong Academy of Agricultural Sciences, China
#
#				Developers: Jinpeng Wang
#############################################################################################################################
###############

#*********************************************************************************FUNCTION usage*****************************************************************************************
# 							This function prints a help message describing all the KARAJ's command-line options							#	 ##########################################################################################################################################################################################																								
# Function name:																						#
#              usage																						#
# ARGUMENTS: 																							#
# 		None																						#
# OUTPUTS: 																							#
# 		Writes String to STDOUT																			#
# RETURN: 																							#
# 		DESCRIPTION OF FLAGS																				#
#*****************************************************************************************************************************************************************************************  
	
	function usage 
		{
			echo ""
			echo "KARAJ:      a command-line software to automate and streamline acquiring biological data"
			echo "Version:    v1.0"
			echo "About:      developed in the BioMedical Machine Learning Lab, University of New South Wales."
			echo "Developers: Mahdieh Labani (co-developer) and Ali Afrasiabi (Co-developer)."
			echo "Code:       https://github.com/mahdieh1/KARAJ"
			echo "Email:      m.labani@unsw.edu.au and a.afrasiabi@unsw.edu.au"
			echo "Citation:"
			echo "Labani M & Afrasiabi A et al, a command-line software to automate and streamline acquiring biological data" 
			echo ""
			echo "Instuction: the list of operations and options that are supported by KARAJ"
			echo ""
			echo "    -l                 list of URL(s), please see exmaples (usage examples -u) or github for further explanation."
			echo "    -p                 list of PMCID(s), please see exmaples (usage examples -u) or github for further explanation."
			echo "    -o                 Output working directory."
			echo "    -t                 type of files: bam/vcf/fastq, please see exmaples (usage examples -u) or github for further explanation."
			echo "    -s                 obtaining suplemenatry data of the corresponding study/studies by specifiying value 1. defult value is 0, which disables the operation."
			echo "    -f                 downloading list of PMCIDs, URLs or accession numbers by passing values 1, 2 and 3, repecteivly."
			echo "                       please see exmaples (usage examples -u) or github for further explanation."
			echo "    -i                 accession number(s): PRJNA/SRP/ERP/GSE/SRR/SRA/SRX/SRS/ERX/ERS/ERP/DRR/DRS/DRX/DRP/GSM/ENCSR/ENCSB/ENCSD/CXR/SAMN."
			echo "    -d                 defualt value is 0 which means downloading data for all accession numbers obtained from URL(s) or PMCID(s)."
			echo "                       by passing value 1 user can select accession numbers to download later on by the summary result."
			echo "    -m                 obtaining metadata table containing sample information and experimental design of the corresponding study."
			echo "    -h                 help."
			echo "    -u                 usage examples."
			echo "    -j                 number of cores."
			echo "    -n                 obtaining processed data of the corresponding study/studies by specifying value 1. de-fault value is 0, which disables the operation."

			exit 1
		}
	
#*********************************************************************************FUNCTION Example****************************************************************************************
# 									This function prints 3 KARAJ commands examples									#	##########################################################################################################################################################################################																								
# Function name:																						#
#              Example																						#
# ARGUMENTS: 																							#
# 		None																						#
# OUTPUTS: 																							#
# 		Writes String to STDOUT																			#
# RETURN: 																							#
# 		Examples of KARAJ commands																			#
#*****************************************************************************************************************************************************************************************  	
    
     function Example
		{
		 echo "Example of how to use KARAJ.
		 *********************************
		 Example1:
		 Example2:
		 Example3:"
		 exit 1
		}



## specifying the options
while getopts ":t:g:u:h" opt; do
 	 case ${opt} in
		    g)
	          tag_file="$OPTARG"
			  printf "Tag file name is %s \n" $OPTARG
		      ;;
 
		    t)
		      type="$OPTARG"
			  printf "Type is %s\n" $OPTARG
		      ;; 
		      
		    u)
		      Example
		      exit 0
		      ;; 
		      
		    h)
		      usage
		      exit 0
		      ;; 
		      
		    \?)
		      echo "Invalid option, please see the Help using -h option:" >&2
		      exit 0
		      ;;
		      
		    \:) 
		      printf "Argument missing from -%s option\n" $OPTARG 
		      exit 0
		      ;;
   
  	esac
done
shift $((OPTIND -1))

Rscript Src/Script01.get.amPeak.R

perl Src/Script02.get.genotypes.pl $type $tag_file 
